Chat Template Doc Fixes (#40173)
huggingface/transformers: docs/source/en/chat_templating_writing.md

### Tool definitions

> Transformers chat template methods allow a user to pass tools as Python functions or a JSON schema. When functions are passed, a JSON schema is automatically generated and passed to the template. Therefore, when your template accesses the `tools` variable, it will always be list of JSON schemas.

Reviewer: ```suggestion
[Tools](./chat_extras) are passed as Python functions or a JSON schema. When functions are passed, a JSON schema is automatically generated and passed to the template. When a template accesses the `tools` variable, it is always a list of JSON schemas.
```

> Even though your template will always receive tools as JSON schema, you may need to radically change this format when rendering them to match the format your model was trained with. For example, [Command-R](./model_doc/cohere) was trained with tools defined with Python function headers! The template internally converts JSON schema types and renders the input tools as Python headers.
> 
> Here is an example of a tool definition in JSON schema format:

Reviewer: ```suggestion
Even though a template always receive tools as a JSON schema, you may need to radically change this format when rendering them to match the format a model was trained with. For example, [Command-R](./model_doc/cohere) was trained with tools defined with Python function headers. The template internally converts JSON schema types and renders the input tools as Python headers.

The example below shows how a tool is defined in JSON schema format.
```

### Tool calls

> In addition to rendering the tool definitions, you will also need to render tool calls and tool responses in your template.
> 
> Tool calls are generally passed in the `tool_calls` key of an `"assistant”` message. This is always a list even though most tool-calling models only support single tool calls, which means the list usually only contains a single element.

Reviewer: ```suggestion
In addition to rendering the tool definitions, you also need to render **tool calls** and **tool responses** in the template.

Tool calls are generally passed in the `tool_calls` key of an `"assistant”` message. This is always a list even though most tool-calling models only support single tool calls, which means the list usually only contains a single element.
```

}
```

> A common pattern for handling tool calls is shown below, but, again, make sure you actually match the format your model was trained with!

Reviewer: ```suggestion
A common pattern for handling tool calls is shown below. Make sure it actually matches the format a model was trained with.
```

### Tool responses

> Tool responses are message dicts with the `tool` role. They are much simpler than tool calls, and usually only contain `role`, `name` and `content` keys:

Reviewer: ```suggestion
Tool responses are message dicts with the `tool` role. They are much simpler than tool calls, and usually only contain the `role`, `name` and `content` keys.
```

}
```

> Some templates may not even need the `name` key. If this is the case for your model, you can write your template to only read the `content` key:

Reviewer: ```suggestion
Some templates may not even need the `name` key, in which case, you can write your template to only read the `content` key.
```

## Contribute

> Once your template is ready, you can use it by setting it as the `chat_template` attribute in the tokenizer and testing it with [`~PreTrainedTokenizerBase.apply_chat_template`]. If it works as expected, then you can upload it to the Hub with with [`~PreTrainedTokenizer.push_to_hub`].

Reviewer: ```suggestion
Once a template is ready, set it with the `chat_template` attribute in the tokenizer and test it with [`~PreTrainedTokenizerBase.apply_chat_template`]. If it works as expected, then upload it to the Hub with [`~PreTrainedTokenizer.push_to_hub`].
```
